### PR TITLE
Better error when DFS returns an empty list of features

### LINKED ISSUE #1309: Raise better error when DFS returns an empty list of features
#### Bug/Feature Request Description
With some combinations of data and primitives, `DeepFeatureSynthesis.build_features()` can return an empty list of features. When this empty list is subsequently passed to `calculate_feature_matrix` an error is raised: `AssertionError: features must be a non-empty list of features`. This error can be confusing for users as they have not specified the features manually in this scenario.

We could catch this situation in the `dfs` function prior to calling `calculate_feature_matrix` and raise a more meaningful error so that users can better understand what path to take to resolve the issue.

#### Expected Output
A more meaningful error should be raise, perhaps something along the lines of "No features can be generated from the specified primitives. Please make sure the primitives you are using are compatible with the variable types in your data."

#### Code to reproduce current AssertionError:
```python
import featuretools as ft
import pandas as pd

data = ft.demo.load_mock_customer(return_entityset=False, n_customers=5)
# Drop zip_code to prevent `zip_code` IdentityFeature from being created
data['customers'] = data['customers'].drop(columns='zip_code')

es = ft.EntitySet()
es.entity_from_dataframe(entity_id='customers', dataframe=data['customers'], index='customer_id', time_index='join_date')
es.entity_from_dataframe(entity_id='sessions', dataframe=data['sessions'], index='session_id', time_index='session_start')
es.add_relationship(ft.Relationship(es['customers']['customer_id'], es['sessions']['customer_id']))

ft.dfs(entityset=es, target_entity='customers', trans_primitives=[], agg_primitives=['max'])
```

### PR DESCRIPTION
**Pull request by DavidQi**
- Fixes #1309

### CODE REVIEW

File: featuretools/computational_backends/calculate_feature_matrix.py [hunk 1]
```diff
@@ -147,7 +147,7 @@ def calculate_feature_matrix(features, entityset=None, cutoff_time=None, instanc
     """
     assert (isinstance(features, list) and features != [] and
             all([isinstance(feature, FeatureBase) for feature in features])), \
-        "features must be a non-empty list of features"
+        "No features can be generated from the specified primitives. Please make sure the primitives you are using are compatible with the variable types in your data."
```
[the remark this answers is not in the document]
Comment (mingdavidqi) [reply featuretools/computational_backends/calculate_feature_matrix.py]: moved it from calculate_feature_matrix to dfs

File: featuretools/synthesis/dfs.py [hunk 2]
```diff
@@ -261,6 +262,10 @@ def dfs(dataframes=None,
     features = dfs_object.build_features(
         verbose=verbose, return_types=return_types)
 
+    assert (isinstance(features, list) and features != [] and
+            all([isinstance(feature, FeatureBase) for feature in features])), \
+        "No features can be generated from the specified primitives. Please make sure the primitives you are using are compatible with the variable types in your data."
+
```
Comment (rwedge) [inline featuretools/synthesis/dfs.py:268]: Let's move this check below where a `features_only` run of `dfs` would return so this check does not cause a `features_only` run of `dfs` to error if no features were created

File: featuretools/synthesis/dfs.py [hunk 3]
```diff
@@ -261,6 +262,10 @@ def dfs(dataframes=None,
     features = dfs_object.build_features(
         verbose=verbose, return_types=return_types)
 
+    assert (isinstance(features, list) and features != [] and
+            all([isinstance(feature, FeatureBase) for feature in features])), \
```
Comment (rwedge) [inline featuretools/synthesis/dfs.py:266]: let's limit this to checking for an empty list

Comment (mingdavidqi) [reply featuretools/synthesis/dfs.py]: Will to move the checking after feature_only process resolve this?

Comment (rwedge) [reply featuretools/synthesis/dfs.py]: No, I'd also like to simplify the assert -- I'd like to tie this specific error message to the feature list being empty.  The other two situations will get caught by the check in `calculate_feature_matrix` , and would suggest some pretty strange things going on that might not fit this new error message

Comment (mingdavidqi) [reply featuretools/synthesis/dfs.py]: updated to assert only with features != []

File: featuretools/tests/synthesis/test_dfs_method.py [hunk 4]
```diff
@@ -53,6 +56,44 @@ def datetime_es():
...
+        })
+        games = pd.DataFrame({
+            'id': range(5),
+            'home_team_id': [2, 2, 1, 0, 1],
+            'away_team_id': [1, 0, 2, 1, 0],
+            'home_team_score': [3, 0, 1, 0, 4],
+            'away_team_score': [2, 1, 2, 0, 0]
+        })
+        dataframes = {'teams': (teams, 'id', None, {'name': 'natural_language'}), 'games': (games, 'id')}
+        relationships = [('teams', 'id', 'games', 'home_team_id')]
```
Comment (rwedge) [inline featuretools/tests/synthesis/test_dfs_method.py:75]: Let's move the lines defining the dataset outside of the `with` block since they aren't going to error and don't need patching

File: featuretools/tests/synthesis/test_dfs_method.py [hunk 5]
```diff
@@ -53,6 +56,44 @@ def datetime_es():
...
+        dfs(dataframes, relationships, target_dataframe_name="teams", features_only=False)
+
+
+def test_dfs_empty_features_features_only():
```
Comment (rwedge) [inline featuretools/tests/synthesis/test_dfs_method.py:79]: instead of a second test, what if the first test had a structure like this

```
with patch(DFSObject):
    test features_only=True returns empty list
    with pytest.raises(AssertionError):
        test features_only=False raises expected error message
```

that way we'd test if we can return an empty list of features in the features_only case

File: featuretools/tests/synthesis/test_dfs_method.py (hunk 4 above)
Comment (mingdavidqi) [reply featuretools/tests/synthesis/test_dfs_method.py]: moved out of with statement

File: featuretools/tests/synthesis/test_dfs_method.py (hunk 5 above)
Comment (mingdavidqi) [reply featuretools/tests/synthesis/test_dfs_method.py]: combined test cases for either feature_only or not

File: featuretools/tests/synthesis/test_dfs_method.py [hunk 6]
```diff
@@ -53,6 +56,28 @@ def datetime_es():
...
+        'id': range(5),
+        'home_team_id': [2, 2, 1, 0, 1],
+        'away_team_id': [1, 0, 2, 1, 0],
+        'home_team_score': [3, 0, 1, 0, 4],
+        'away_team_score': [2, 1, 2, 0, 0]
+    })
+    dataframes = {'teams': (teams, 'id', None, {'name': 'natural_language'}), 'games': (games, 'id')}
+    relationships = [('teams', 'id', 'games', 'home_team_id')]
+    features = dfs(dataframes, relationships, target_dataframe_name="teams", features_only=True)
+    assert (isinstance(features, list) and features != []), True
```
Comment (rwedge) [inline featuretools/tests/synthesis/test_dfs_method.py:75]: I'd like the patch of `DeepFeatureSynthesis` returning `[]` to also apply to this run dfs / assert code so we can test that an empty feature list is fine if `features_only=True`

we'd need to separate the `with` block into two blocks: the patch context which would apply to both `dfs` calls, and the `pytest.raises` with block only applying to the `features_only=False` call

Comment (mingdavidqi) [reply featuretools/tests/synthesis/test_dfs_method.py]: patch DeepFeatureSynthesis to return []

**Review by rwedge — APPROVED** [review]
Looks good!  Congrats on your first contribution to Featuretools!
